sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/objects.scala

```diff
@@ -228,3 +228,31 @@ object ObjectSerializerPruning extends Rule[LogicalPlan] {
       }
   }
 }
+
+/**
+ * Reassigns per-query unique IDs to `LambdaVariable`s, whose original IDs are globally unique. This
+ * can help Spark to hit codegen cache more often and improve performance.
+ */
+object ReassignLambdaVariableID extends Rule[LogicalPlan] {
+  override def apply(plan: LogicalPlan): LogicalPlan = {
+    if (!SQLConf.get.getConf(SQLConf.OPTIMIZER_REASSIGN_LAMBDA_VARIABLE_ID)) return plan
+
+    // The original LambdaVariable IDs are all positive. To avoid conflicts, the new IDs are all
+    // negative and starts with -1.
+    var newId = -1L
+    val oldIdToNewId = scala.collection.mutable.Map.empty[Long, Long]
+
+    plan.transformAllExpressions {
```

Comment:
The two traversals on the plan here sure makes the intent clean: one pass for collecting old-to-new ID mappings, and then another pass to actually do the transformation.

But efficiency-wise, these two traversals can be combined into one easily, right? Reducing the number of traversals can help save a lot of time when dealing with large plans.

Reply:
fixed.